src/ipc/ipc4/mtrace.c

```diff
@@ -0,0 +1,170 @@
+#include <zephyr/logging/log.h>
+
+#include <sof/schedule/edf_schedule.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+LOG_MODULE_REGISTER(mtrace, CONFIG_SOF_LOG_LEVEL);
+
+/**
+ * If mtrace log buffer has less space than
+ * the threshold, notify host with BUFFER_STATUS IPC4.
+ */
+#define NOTIFY_BUFFER_STATUS_THRESHOLD 2048
+
+/**
+ * Default aging timer value. This time defines the maximum time to
+ * block IPC BUFFER STATUS notify messages. The notification is sent
+ * either due to amount of data available in the buffer, or by timeout
+ * in case logs are sent at slow rate.
+ */
+#define IPC4_MTRACE_NOTIFY_AGING_TIMER_MS 1000
+
+/**
+ * Smallest possible aging timer value
+ */
+#define IPC4_MTRACE_AGING_TIMER_MIN_MS 100
+
+/* bb2aa22e-1ab6-4650-8501-6e67fcc04f4e */
+DECLARE_SOF_UUID("mtrace-task", mtrace_task_uuid, 0xbb2aa22e, 0x1ab6, 0x4650,
+		 0x85, 0x01, 0x6e, 0x67, 0xfc, 0xc0, 0x4f, 0x4e);
+
+static uint64_t mtrace_notify_last_sent;
+
+static uint32_t mtrace_bytes_pending;
+
+static uint32_t mtrace_aging_timer = IPC4_MTRACE_NOTIFY_AGING_TIMER_MS;
+
+struct task mtrace_task;
+
+#define MTRACE_IPC_CORE	PLATFORM_PRIMARY_CORE_ID
+
+static void mtrace_log_hook(size_t written, size_t space_left)
+{
+	if (arch_proc_id() == MTRACE_IPC_CORE &&
+	    (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	     k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer)) {
+		ipc_send_buffer_status_notify();
+		mtrace_notify_last_sent = k_uptime_get();
+		mtrace_bytes_pending = 0;
+
+	} else {
+		mtrace_bytes_pending += written;
+
+		/* TODO: if hook is called on non-zero core, logs
+		 * maybe lost with too slow aging timer. need to
+		 * figure out a safe way to wake up the mtrace task
+		 * from another core.
+		 */
+	}
+}
+
+static enum task_state mtrace_task_run(void *data)
+{
+	if (k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer &&
+	    mtrace_bytes_pending)
+		mtrace_log_hook(0, 0);
+
+	/* task will be re-run based on mtrace_task_deadline() */
+	return SOF_TASK_STATE_RESCHEDULE;
+}
+
+static uint64_t mtrace_task_deadline(void *data)
+{
+	return k_uptime_ticks() + k_ms_to_ticks_ceil64(mtrace_aging_timer);
+}
+
+int ipc4_set_enable_logs(bool first_block,
+			 bool last_block,
+			 uint32_t data_offset_or_size,
+			 char *data)
+{
+	const struct log_backend *log_backend = log_backend_adsp_mtrace_get();
+	struct ipc4_log_state_info *log_state;
+	struct task_ops ops = {
+		.run = mtrace_task_run,
+		.get_deadline = mtrace_task_deadline,
+		.complete = NULL,
+	};
+
+	if (!(first_block && last_block)) {
+		LOG_ERR("log_state data is expected to be sent as one chunk");
+		return -EINVAL;
+	}
+
+	if (data_offset_or_size < sizeof(struct ipc4_log_state_info)) {
+		LOG_ERR("log_state too small data size: %u", data_offset_or_size);
+		return -EINVAL;
+	}
+
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)data, data_offset_or_size);
+
+	log_state = (struct ipc4_log_state_info *)data;
```

Comment:
Actually this is not correct, for memory window log type this IPC will be extended and there is tracing_method information which we should use and then decide to run such mtrace init because in current form this will also trigger for probe log enable

Reply:
Ack @abonislawski . I enabled only the basic set as driver does not supported the extended struct yet.  FYI @ujfalusi we need to extend the driver side if we use the ENABLE_LOGS IPC to also enable the probes logging.

Reply:
@abonislawski ^^ I left a TODO to code for this, I don't have anything to test this against currently, so I prefer to start with something I can test.

Reply:
Thats absolutely fine, I can extend this later as I can easily test both types of this ipc message but please extend this in driver because extended message version is actually used for mtrace and normal (this one) version for probes :)